benches/latency.rs

```diff
@@ -0,0 +1,218 @@
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<(), io::Error> {
+        loop {
+            if self.stop.load(Ordering::SeqCst) {
+                return Ok(futures::Async::Ready(()))
+            }
+
+            if let Some((size, peer)) = self.to_send.take() {
+                match self.socket.send_to(&self.buf[..size], &peer) {
+                    Err(e) => {
+                        self.to_send = Some((size, peer));
+                        return try_nb!(Err(e));
+                    },
+                    Ok(_) => { }
+                }
+            }
+
+            self.to_send = Some(
+                try_nb!(self.socket.recv_from(&mut self.buf))
+                );
+        }
+    }
+}
+
+/// UDP echo server
+struct ChanServer {
+    rx : tokio_core::channel::Receiver<usize>,
+    tx : tokio_core::channel::Sender<usize>,
+    buf : Option<usize>,
+    stop : Arc<AtomicBool>,
+}
+
+impl ChanServer {
+    fn new(tx : tokio_core::channel::Sender<usize>, rx : tokio_core::channel::Receiver<usize>, stop : Arc<AtomicBool>) -> Self {
+        ChanServer {
+            rx: rx,
+            tx: tx,
+            buf: None,
+            stop: stop,
+        }
+    }
+}
+
+impl Future for ChanServer {
+    type Item = ();
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<(), io::Error> {
+        loop {
+            if self.stop.load(Ordering::SeqCst) {
+                return Ok(futures::Async::Ready(()))
+            }
+
+            if let Some(u) = self.buf.take() {
+                match self.tx.send(u) {
+                    Err(e) => {
+                        self.buf = Some(u);
+                        return try_nb!(Err(e));
+                    },
+                    Ok(_) => { }
+                }
+            }
+
+            match self.rx.poll() {
+                Ok(futures::Async::Ready(None)) => return Ok(futures::Async::Ready(())),
+                Ok(futures::Async::Ready(Some(t))) => { self.buf = Some(t) },
+                Ok(futures::Async::NotReady) => return Ok(futures::Async::NotReady),
+                Err(e) => return try_nb!(Err(e)),
+            }
+        }
+    }
+}
+#[bench]
+fn udp_echo_latency(b: &mut Bencher) {
+    let server_addr= "127.0.0.1:7398".to_string();
+    let server_addr = server_addr.parse::<SocketAddr>().unwrap();
+    let client_addr= "127.0.0.1:7399".to_string();
+    let client_addr = client_addr.parse::<SocketAddr>().unwrap();
+
+    let stop = Arc::new(AtomicBool::new(false));
+    let stop2 = stop.clone();
+
+    let child = thread::spawn(move || {
+        let mut l = Core::new().unwrap();
+        let handle = l.handle();
+
+        let socket = tokio_core::net::UdpSocket::bind(&server_addr, &handle).unwrap();
+
+        let server = EchoServer::new(socket, stop);
+
+        l.run(server).unwrap();
+    });
+
+    // TODO: More reliable way to bind server socket and start server
+    // first
+    thread::sleep(Duration::from_millis(100));
```

Comment:
Could a channel be used instead of this to send a message as soon as the socket is bound?